odoo/tools/safe_eval.py

```diff
@@ -1,225 +1,812 @@
-    """test_expr(expression, allowed_codes[, mode[, filename]]) -> code_object
+        We uses a white-list approach but with some exceptions that are checked in the function.:
+            - The safe_eval function
+            - Objects that their __self__ is an unsafe module (We can prevent functions of the like of os.system that way)
+                Note : Objects present in _BUILTINS are not checked, as they are already safe
+
+        Raises:
+            ValueError: If the object is not safe
+
+        Args:
+            __obj (Any): The object to be checked
+        """
 
-    Test that the expression contains only the allowed opcodes.
-    If the expression is valid and contains only allowed codes,
-    return the compiled code object.
-    Otherwise raise a ValueError, a Syntax Error or TypeError accordingly.
+        obj = args.pop("__obj")
+
+        # The black-list part (See docstring)
+        if obj is safe_eval or (
+            callable(obj)
+            and not isinstance(obj, SafeWrapper)
+            and hasattr(obj, "__self__")
+            and isinstance(obj.__self__, types.ModuleType)
+            and obj.__self__.__name__ not in _ALLOWED_MODULES
+            and obj not in _BUILTINS.values()
+        ):
+            raise ValueError(f"unsafe object: {obj} (denied by black-list)")
+
+        if isinstance(obj, allowed_instance) or (
+            isinstance(obj, type) and issubclass(obj, allowed_type)
+        ):
+            return obj
+        elif (
+            isinstance(obj, _wrapped_instances)
+            or (isinstance(obj, type) and issubclass(obj, _wrapped_types))
+            or (isinstance(obj, types.ModuleType) and obj.__name__ in _ALLOWED_MODULES)
+        ):
+            return SafeWrapper(__obj=obj, __type_checker=type_checker)
+        else:
+            raise ValueError(f"unsafe object: {obj} (type not present in white-list)")
+
+    def call_checker(**kwargs) -> Any:
+        """Function that checks if a given function and its arguments are safe.
+
+        This function will work with the type checker to check if a given function and its arguments are safe.
+
+        Args:
+            __func (Callable): The function to be checked
+            __args (tuple[Any]): The arguments of the function
+            __kwargs (dict[str, Any]): The keyword arguments of the function
+
+        Returns:
+            Any: Returns the value returned by the function if it's safe.
+                    Or it will be wrapped with SafeWrapper if it's not safe.
+                    Otherwise, an exception will be raised from the type checker.
+        """
 
-    :param filename: optional pseudo-filename for the compiled expression,
-                 displayed for example in traceback frames
-    :type filename: string
+        func = kwargs.pop("__func")
+        f_args = kwargs.pop("__args")
+        f_kwargs = kwargs.pop("__kwargs")
+
+        for arg in (f_args, f_kwargs.values()):
+            type_checker(__obj=arg)
```

Comment:
`arg` is a list at this point no ?

it should be `for arg in (*f_args, *f_kwargs.values()):` I think